decorate_test.go

```diff
@@ -0,0 +1,463 @@
+		child := c.Scope("child")
+		require.NoError(t, child.Provide(func() *A { return &A{name: "A"} }, dig.Export(true)))
+
+		assert.NoError(t, child.Decorate(func(a *A) *A { return &A{name: a.name + "'"} }))
+		assert.NoError(t, c.Invoke(func(a *A) {
+			assert.Equal(t, "A", a.name, "expected name to equal original name in parent scope")
+		}))
+		assert.NoError(t, child.Invoke(func(a *A) {
+			assert.Equal(t, "A'", a.name, "expected name to equal decorated name in child scope")
+		}))
+	})
+
+	t.Run("simple decorate a provider to a scope and its descendants", func(t *testing.T) {
+		t.Parallel()
+		type A struct {
+			name string
+		}
+
+		c := dig.New()
+		child := c.Scope("child")
+		require.NoError(t, c.Provide(func() *A { return &A{name: "A"} }))
+
+		assert.NoError(t, c.Decorate(func(a *A) *A { return &A{name: a.name + "'"} }))
+		assertDecoratedName := func(a *A) {
+			assert.Equal(t, a.name, "A'", "expected name to equal decorated name")
+		}
+		assert.NoError(t, c.Invoke(assertDecoratedName))
+		assert.NoError(t, child.Invoke(assertDecoratedName))
+	})
+
+	t.Run("modifications compose with descendants", func(t *testing.T) {
+		t.Parallel()
+		type A struct {
+			name string
+		}
+
+		c := dig.New()
+		child := c.Scope("child")
+		require.NoError(t, c.Provide(func() *A { return &A{name: "A"} }))
+
+		require.NoError(t, c.Decorate(func(a *A) *A { return &A{name: a.name + "'"} }))
+		require.NoError(t, child.Decorate(func(a *A) *A { return &A{name: a.name + "'"} }))
+
+		assert.NoError(t, c.Invoke(func(a *A) {
+			assert.Equal(t, "A'", a.name, "expected decorated name in parent")
+		}))
+
+		assert.NoError(t, child.Invoke(func(a *A) {
+			assert.Equal(t, "A''", a.name, "expected double-decorated name in child")
+		}))
+
+		sibling := c.Scope("sibling")
+		grandchild := child.Scope("grandchild")
+		sibling.Invoke(func(a *A) {
+			assert.Equal(t, "A'", a.name, "expected single-decorated name in sibling")
+		})
+		grandchild.Invoke(func(a *A) {
+			assert.Equal(t, "A''", a.name, "expected double-decorated name in grandchild")
+		})
+	})
+
+	t.Run("decorate with In struct", func(t *testing.T) {
+		t.Parallel()
+
+		type A struct {
+			Name string
+		}
+		type B struct {
+			dig.In
+
+			A *A
+			B string `name:"b"`
+		}
+
+		type C struct {
+			dig.Out
+
+			A *A
+			B string `name:"b"`
+		}
+
+		c := dig.New()
+		require.NoError(t, c.Provide(func() *A { return &A{Name: "A"} }))
+		require.NoError(t, c.Provide(func() string { return "b" }, dig.Name("b")))
+
+		require.NoError(t, c.Decorate(func(b B) C {
+			return C{
+				A: &A{
+					Name: b.A.Name + "'",
+				},
+				B: b.B + "'",
+			}
+		}))
+
+		assert.NoError(t, c.Invoke(func(b B) {
+			assert.Equal(t, "A'", b.A.Name)
+			assert.Equal(t, "b'", b.B)
+		}))
+	})
+
+	t.Run("decorate with value groups", func(t *testing.T) {
+		type Params struct {
+			dig.In
+
+			Animals []string `group:"animals"`
+		}
+
+		type Result struct {
+			dig.Out
+
+			Animals []string `group:"animals"`
+		}
```

Comment:
This is pretty cool.
We'll need to document under Decorate that for value groups, the result must be a `[]T` as well.